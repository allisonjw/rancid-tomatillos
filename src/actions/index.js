export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies
});

export const getRatings = ratings => ({
  type: 'GET_RATINGS',
  ratings
});

export const addRating = rating => ({
  type: 'ADD_RATING',
  rating
});

export const getUser = user => ({
  type: 'ADD_USER',
  user
});

export const deleteRating = rating => ({
  type: 'REMOVE_RATING',
  rating
});

export const logOut = () => ({
  type: 'LOG_OUT',
})

export const setError = (message) => ({
  type: 'HAS_ERROR',
  message
})