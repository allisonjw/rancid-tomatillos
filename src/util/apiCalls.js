export const getMovieData = async (url) => {
    const response = await fetch(url)
    const data = await response.json()
    return data;
};


export const fetchUser = async (email, password) => {
    const userLogin = { email: `${email}`, password: `${password}`};
    const options = {
        method: 'POST',
        body: JSON.stringify(userLogin),
        headers: {
            'Content-Type': 'application/json'
        }
    };

    const response = await fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    if(!response.ok){
        throw new Error('oops! please check your username and password are correct. ')
    }
    return response.json();
};

export const postRating = async (userRating, userID) => {
  const options = {
      method: 'POST',
      body: JSON.stringify(userRating),
      headers: {
          'Content-Type': 'application/json'
      }
  };

  const post = await fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userID}/ratings`, options)
  const rating = await post.json()
  return rating;
};

export const fetchRatings = userID => {
    return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userID}/ratings`).then(response => {
      if (!response.ok) {
        throw Error('There was a problem fetching all ratings');
      }
      return response.json();
    });
};


export const removeRating = async (userID, ratingID) => {
    const url = `https://the-screening-room-db.herokuapp.com/api/v1/users/${userID}/moviefavorites/${ratingID}`;
        const options = {
            method: "DELETE",
            headers: {
                "Content-Type": "application/json"
            }
        };
        const response = await fetch(url, options);
        if (!response.ok) {
            throw Error('There was changing your rate')
        }
        return response.json()

};
