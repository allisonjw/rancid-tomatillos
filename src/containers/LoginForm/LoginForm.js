import React, { Component } from 'react';
import '../App/App.scss';
import { Redirect } from 'react-router-dom';
import { fetchUser } from '../../util/apiCalls';
import { connect } from 'react-redux';
import { getUser } from '../../actions/index'


export class LoginForm extends Component {
  constructor() {
    super()
    this.state = {
      email: '',
      password: '', 
      loggedIn: false,
      isPasswordShown: false
    }
  }
  
  togglePasswordVisibility = () => {
    const { isPasswordShown } = this.state;
    this.setState({ isPasswordShown: !isPasswordShown})
  }

  handleChange = e => {
    this.setState({[e.target.name] : e.target.value })
  }

  handleSubmit = (e) => {
    const { email, password } = this.state
    e.preventDefault()
    fetchUser(email, password)
    .then(user => this.props.getUser(user))
  }
    // if(this.state.loggedIn === true){
    //   return <Redirect to="/movies" />;
    // }

  render() {
    const { isPasswordShown } = this.state;
    return (
      <section className="section-form">
        <div className="section-form__header">
        <div className="user-info">
          <div className="user-info__form">
            <form className="loginForm">
              <h1 className="heading">
                Please login!
              </h1>
              <div className="loginForm__group">
                <input 
                  name="email"
                  value={this.state.email}
                  className="loginForm__input"
                  onChange={(e)=> this.handleChange(e)}
                  id="email"
                  placeholder="email"
                  type="text"
                  />
                <label 
                  for="email"
                  className="loginForm__label">email
                </label>
                <div className="loginForm__group">
                  <i
                    className={`fa ${isPasswordShown ? "fa-eye-slash" : "fa-eye"} password-icon`} onClick={this.togglePasswordVisibility} />
                  <input
                    name="password"
                    value={this.state.password}
                    className="loginForm__input"
                    onChange={(e) => this.handleChange(e)}
                    id="password"
                    placeholder="password"
                    type={(isPasswordShown) ? "text" : "password"}
                  />
                  <label
                    for="password"
                    className="loginForm__label">password
                  </label>
                  <a className="btn btn-grey" href='/login' onClick={(e)=> this.handleSubmit(e)}>Login &rarr;</a>
                </div>
              </div>
            </form>
          </div>
        </div>
        </div>
      </section>
      
    )
  }
}

const mapDispatchToProps = dispatch => ({
  getUser: user => dispatch(getUser(user))
})

export default connect(null, mapDispatchToProps)(LoginForm)

