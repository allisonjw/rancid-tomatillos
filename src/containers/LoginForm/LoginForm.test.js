import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { getUser } from '../../actions/index';
import { LoginForm, mapDispatchToProps } from './LoginForm';
import { Provider } from 'react-redux';

import { fetchUser } from '../../util/apiCalls';

jest.mock('../../util/apiCalls')

describe('LoginForm', () => {
  let wrapper;
  let mockEvent
  let mockEvent2

  beforeEach(() => {

    wrapper = shallow(
        <LoginForm
        getUser={jest.fn()}
        />
    )

    mockEvent = {target: {name: 'email', value: 'diana@turing.io'}, preventDefault: jest.fn()}
    mockEvent2 = {target: {name: 'password', value: '111111'}, preventDefault: jest.fn()}
  })

  it('should render with correct data', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render with correct data when isPasswordShown is true', () => {
    wrapper.setState({isPasswordShown: true})

    expect(wrapper).toMatchSnapshot()
  })

  it('should render with correct data when loggedIn is true', () => {
    wrapper.setState({loggedIn: true})

    expect(wrapper).toMatchSnapshot()
  })

  it('should call handleChange when input value is changed', () => {

    wrapper.instance().handleChange = jest.fn()

    wrapper.find("#email").simulate('change', mockEvent)

    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockEvent)

    wrapper.find("#password").simulate('change', mockEvent2)

    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockEvent2)

  })

  it('should change state when handleChange is called', () => {

    wrapper.setState({email: ''})

    wrapper.instance().handleChange(mockEvent)

    expect(wrapper.state('email')).toEqual(mockEvent.target.value)
  })

  it('should call handleSubmit when button is clicked', () => {

    wrapper.instance().handleSubmit = jest.fn()

    wrapper.find("a").simulate('click', {preventDefault: jest.fn()})

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled()

  })

  it('should call getUser when handleSubmit runs', async () => {
    const mockUser = {email: 'diane@turing.io', name: 'Diane', id: 7}
    fetchUser.mockImplementation(() => {
      return Promise.resolve(mockUser)
    })
    const mockGetUser = jest.fn()
    const wrapper = shallow(
        <LoginForm
        getUser={mockGetUser}
        />
    )

    await wrapper.instance().handleSubmit(mockEvent);

    expect(mockGetUser).toHaveBeenCalledWith(mockUser)
  })

  it('should set loggedIn to be true when handleSubmit runs', async () => {
    const mockUser = {email: 'diane@turing.io', name: 'Diane', id: 7}

    fetchUser.mockImplementation(() => {
      return Promise.resolve(mockUser)
    })

    const mockGetUser = jest.fn()

    expect(wrapper.state('loggedIn')).toEqual(false)

    await wrapper.instance().handleSubmit(mockEvent);

    expect(wrapper.state('loggedIn')).toEqual(true)
  })

  it('should toggle isPasswordShown in state when togglePasswordVisibility is called', () => {
    wrapper.setState({isPasswordShown: false})

    wrapper.instance().togglePasswordVisibility()

    expect(wrapper.state('isPasswordShown')).toEqual(true)

    wrapper.instance().togglePasswordVisibility()

    expect(wrapper.state('isPasswordShown')).toEqual(false)
  })

  it('should call togglePasswordVisibility when button is clicked', () => {

    wrapper.instance().togglePasswordVisibility = jest.fn()
    wrapper.instance().forceUpdate();

    wrapper.find("#pwIconFirstHalf").simulate('click')

    expect(wrapper.instance().togglePasswordVisibility).toHaveBeenCalled()

    wrapper.find("#pwIconSecondHalf").simulate('click')

    expect(wrapper.instance().togglePasswordVisibility).toHaveBeenCalled()

  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the correct action when getUser is called', () => {
      const mockDispatch = jest.fn();
      const mockUser = {email: 'diane@turing.io', name: 'Diane', id: 7}
      const dispatchedAction = getUser(mockUser);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.getUser(mockUser)

      expect(mockDispatch).toHaveBeenCalledWith(dispatchedAction)
    })
  })
})
