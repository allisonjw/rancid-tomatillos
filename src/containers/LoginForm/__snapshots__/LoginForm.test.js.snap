// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render with correct data 1`] = `
<section
  className="section-form"
>
  <div
    className="section-form__header"
  >
    <div
      className="user-info"
    >
      <div
        className="user-info__form"
      >
        <form
          className="loginForm"
        >
          <h1
            className="heading"
          >
            Please login!
          </h1>
          <div
            className="loginForm__group"
          >
            <input
              className="loginForm__input"
              id="email"
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value=""
            />
            <label
              className="loginForm__label"
              htmlFor="email"
            >
              email
            </label>
            <div
              className="loginForm__group"
            >
              <i
                className="fa fa-eye-slash password-icon"
                onClick={[Function]}
              />
              <input
                className="loginForm__input"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
              />
              <label
                className="loginForm__label"
                htmlFor="password"
              >
                password
              </label>
              <i
                className="fa fa-eye password-icon"
                onClick={[Function]}
              />
              <a
                className="btn btn-grey"
                href="/login"
                onClick={[Function]}
              >
                Login →
              </a>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</section>
`;

exports[`LoginForm should render with correct data when isPasswordShown is true 1`] = `
<section
  className="section-form"
>
  <div
    className="section-form__header"
  >
    <div
      className="user-info"
    >
      <div
        className="user-info__form"
      >
        <form
          className="loginForm"
        >
          <h1
            className="heading"
          >
            Please login!
          </h1>
          <div
            className="loginForm__group"
          >
            <input
              className="loginForm__input"
              id="email"
              name="email"
              onChange={[Function]}
              placeholder="email"
              type="text"
              value=""
            />
            <label
              className="loginForm__label"
              htmlFor="email"
            >
              email
            </label>
            <div
              className="loginForm__group"
            >
              <i
                className="fa fa-eye password-icon"
                onClick={[Function]}
              />
              <input
                className="loginForm__input"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="text"
                value=""
              />
              <label
                className="loginForm__label"
                htmlFor="password"
              >
                password
              </label>
              <i
                className="fa fa-eye-slash password-icon"
                onClick={[Function]}
              />
              <a
                className="btn btn-grey"
                href="/login"
                onClick={[Function]}
              >
                Login →
              </a>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</section>
`;
